### ISSUE TITLE
SerializableTransactionRunner loses records on retry

### ISSUE THREAD

**Issue by sudharsannr** [jdbi/jdbi#1967]
Hi,

I have got a batch insert occuring on a table 'test' with the use of `@SqlBatch` annotation. The batch size is 512 and there are about 1000 records to insert to the table. The configuration for JDBI transaction handler is [SerializableTransactionRunner](https://github.com/jdbi/jdbi/blob/0420cc462f98565062bd577cf5d3a4d92a621060/core/src/main/java/org/jdbi/v3/core/transaction/SerializableTransactionRunner.java).

When a failure occurs from the database during commit especially around [here](https://github.com/jdbi/jdbi/blob/0420cc462f98565062bd577cf5d3a4d92a621060/core/src/main/java/org/jdbi/v3/core/transaction/LocalTransactionHandler.java#L210)
```
 public <R, X extends Exception> R inTransaction(Handle handle,
                                                        HandleCallback<R, X> callback) throws X {
            if (isInTransaction(handle)) {
                throw new IllegalStateException("Already in transaction");
            }
            didTxnRollback = false;
            final R returnValue;
            try {
                handle.begin();
                returnValue = callback.withHandle(handle);
                if (!didTxnRollback) {
                    **// ISSUE: Database throws exception here
                    handle.commit();**
                }
            } catch (Throwable e) {
                try {
                    handle.rollback();
                } catch (Exception rollback) {
                    e.addSuppressed(rollback);
                }
                throw e;
            } finally {
                didTxnRollback = false;
            }

            return returnValue;
        }

```

, the transaction runner tries to re-try the batch around [here](https://github.com/jdbi/jdbi/blob/0420cc462f98565062bd577cf5d3a4d92a621060/core/src/main/java/org/jdbi/v3/core/transaction/SerializableTransactionRunner.java#L53-L79)
```
while (true) {
	try {
               **// Second attempt has 0 bindings and commit transaction completes in 0 ms
		R result = getDelegate().inTransaction(handle, callback);**
		config.onSuccess.accept(new ArrayList<>(failures));
		return result;
	} catch (Exception last) {
		X x = (X) last;

		// throw immediately if the exception is unexpected
		if (!isSqlState(config.serializationFailureSqlState, x)) {
			throw last;
		}

		failures.addLast(x);
		config.onFailure.accept(new ArrayList<>(failures));

		// no more attempts left? Throw ALL the exceptions! \o/
		attempts -= 1;
		if (attempts <= 0) {
			X toThrow = failures.removeLast();
			while (!failures.isEmpty()) {
				toThrow.addSuppressed(failures.removeLast());
			}
			throw toThrow;
		}
	}
}
```

The issue is the retry-attempt will always be successful as the bind is cleared by execute. The definition for [execute](https://github.com/jdbi/jdbi/blob/0420cc462f98565062bd577cf5d3a4d92a621060/sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/SqlBatchHandler.java#L68-L71) is:

`Function<PreparedBatch, ResultIterator<?>> modCounts = PreparedBatch::executeAndGetModCount;
batchIntermediate = method.getReturnType().equals(boolean[].class) ? mapToBoolean(modCounts) : modCounts;`

This particular execution [clears the bind](https://github.com/jdbi/jdbi/blob/127fda14d9875518670a4b568fdd84688a2a95ea/core/src/main/java/org/jdbi/v3/core/statement/PreparedBatch.java#L236) as a result, the subsequent retry doesn't have any bind and is marked successful. 

An example scenario per my batch insert for 1000 records with batch size of 512 is as follows:

1. Batch#1 with 512 records fails from the database and records are not committed.
2. SerializableTransactionRunner retries and is successful with empty binds.
3. Batch#2 with 488 records are successful.

Queried the table and Batch#1 consisting of the 512 records are missing.

Edit:
The issue fix should be around at [SqlBatchTransactionHandler](https://github.com/jdbi/jdbi/blob/0420cc462f98565062bd577cf5d3a4d92a621060/sqlobject/src/main/java/org/jdbi/v3/sqlobject/statement/internal/SqlBatchHandler.java#L311-L319):

```
private ResultIterator<?> executeBatch(final Handle handle, final PreparedBatch batch) {
	if (!handle.isInTransaction() && sqlBatch.transactional()) {
		// it is safe to use same prepared batch as the inTransaction passes in the same
		// Handle instance.
		// Actual:
                // return handle.inTransaction(c -> batchIntermediate.apply(batch));
                // Avoid clearing the original batch with binds by cloning
		return handle.inTransaction(c -> batchIntermediate.apply(deepCopy(batch)));


	} else {
		return batchIntermediate.apply(batch);
	}
}
```

**Comment by hgschmie** [comment]
Thank you for the detailed report. We will look into this (need to write a test case).

**Comment by stevenschlansker** [comment]
Fixed in 3.27.1
